Sources/TuistGenerator/Mappers/AutogeneratedSchemesProjectMapper.swift

```diff
@@ -1,150 +1,323 @@
-        let schemeNames = Set(project.schemes.map(\.name))
-        let schemes = project.schemes
-
-        let autogeneratedSchemes = project.targets.compactMap { (target: Target) -> Scheme? in
-            let scheme = self.createDefaultScheme(
-                target: target,
-                project: project,
-                buildConfiguration: project.defaultDebugBuildConfigurationName
+        let userDefinedSchemes = project.schemes
+        let userDefinedSchemeNames = Set(project.schemes.map(\.name))
+
+        var buildTargets: Set<Target> = []
+        var testTargets: Set<Target> = []
+        var runTargets: Set<Target> = []
+        project.targets.forEach { target in
+            switch target.product {
+            case .app, .appClip, .commandLineTool, .watch2App:
+                runTargets.insert(target)
+            case .uiTests, .unitTests:
+                testTargets.insert(target)
+            case .framework, .staticLibrary, .dynamicLibrary, .staticFramework, .bundle, .appExtension, .watch2Extension,
+                 .tvTopShelfExtension, .messagesExtension, .stickerPackExtension:
+                buildTargets.insert(target)
+            }
+        }
+
+        let targetGroups: [String: (build: Set<Target>, test: Set<Target>, run: Set<Target>)]
+        switch project.options.targetSchemesGrouping {
+        case .singleScheme:
+            targetGroups = [
+                project.name: (
+                    build: buildTargets,
+                    test: testTargets,
+                    run: runTargets
+                ),
+            ]
+        case let .byNameSuffix(buildSuffixes, testSuffixes, runSuffixes):
+            targetGroups = groupByName(
+                buildTargets: &buildTargets,
+                testTargets: &testTargets,
+                runTargets: &runTargets,
+                buildSuffixes: buildSuffixes,
+                testSuffixes: testSuffixes,
+                runSuffixes: runSuffixes
             )
-            // The user has already defined a scheme with that name.
-            if schemeNames.contains(scheme.name) { return nil }
-            return scheme
+        case .notGrouped:
+            targetGroups = [:]
+        case .none:
+            return (project, [])
         }
 
-        return (project.with(schemes: schemes + autogeneratedSchemes), [])
+        let autogeneratedSchemes: [Scheme]
+        if !targetGroups.isEmpty {
+            autogeneratedSchemes = targetGroups.map { name, targets in
+                createDefaultScheme(
+                    name: name,
+                    project: project,
+                    buildTargets: targets.build,
+                    testTargets: targets.test,
+                    runTargets: targets.run,
+                    buildConfiguration: project.defaultDebugBuildConfigurationName
+                )
+            }
+        } else {
+            let remainingBuildTargetsSchemes = buildTargets.map {
```

Comment:
There's sadly a regression when using the `.notGrouped` code-path for extensions 😞

The host app for extensions isn't being added to the list of build targets, once you open Xcode's scheme editor - Xcode will modify the scheme automatically to add the host app (resulting in unstable generated schemes 🤐)

Reply:
isn't it added by the method that generrates the scheme? 🤔